modules/core/src/main/java/org/apache/ignite/spi/deployment/local/LocalDeploymentSpi.java

```diff
@@ -171,38 +237,33 @@ private String getResourceName(String clsName, Map<String, String> rsrcs) {
     @Override public boolean register(ClassLoader ldr, Class<?> rsrc) throws IgniteSpiException {
         assert ldr != null;
         assert rsrc != null;
+        boolean move = false;
 
         if (log.isDebugEnabled())
             log.debug("Registering [ldrRsrcs=" + ldrRsrcs + ", ldr=" + ldr + ", rsrc=" + rsrc + ']');
 
-        ConcurrentMap<String, String> clsLdrRsrcs = ldrRsrcs.getSafe(ldr);
+        Map<String, String> newRsrcs;
 
-        if (clsLdrRsrcs == null) {
-            ConcurrentMap<String, String> old = ldrRsrcs.putIfAbsent(ldr,
-                clsLdrRsrcs = new ConcurrentHashMap<>());
-
-            if (old != null)
-                clsLdrRsrcs = old;
-        }
+        synchronized (this) {
+            ConcurrentMap<String, String> clsLdrRsrcs = ldrRsrcs.getSafe(ldr);
 
-        Map<String, String> newRsrcs = addResource(ldr, clsLdrRsrcs, rsrc);
+            // move forward, localDeployTask compatibility issue.
+            if (clsLdrRsrcs != null && ldrRsrcs.size() > 1) {
+                ldrRsrcs.remove(ldr);
 
-        if (!getBoolean(IGNITE_DEPLOYMENT_PRESERVE_LOCAL)) {
-            Collection<ClassLoader> rmvClsLdrs = null;
+                move = true;
+            }
 
-            if (!F.isEmpty(newRsrcs)) {
-                rmvClsLdrs = new LinkedList<>();
+            ConcurrentMap<String, String> old = ldrRsrcs.putIfAbsent(ldr,
```

Comment:
Try to use a ldrRsrcs.compute method here.
It may get rid of synchronization and change order in the map in that places where it requires.

Reply:
don`t understand how .compute will help to move forward here.

Reply:
ok, i fix it.